flink-runtime/src/test/java/org/apache/flink/runtime/testutils/InMemoryReporter.java

```diff
@@ -159,17 +168,51 @@ void applyRemovals() {
 
     @Override
     public void notifyOfAddedMetric(Metric metric, String metricName, MetricGroup group) {
-        MetricGroup metricGroup = unwrap(group);
-        LOG.debug("Registered {} @ {}", metricName, metricGroup);
-        synchronized (this) {
-            metrics.computeIfAbsent(metricGroup, dummy -> new HashMap<>()).put(metricName, metric);
+        if (captureMode != CaptureMode.NONE) {
+            MetricGroup metricGroup = unwrap(group);
+            LOG.debug("Registered {} @ {}", metricName, metricGroup);
+            synchronized (this) {
+                metrics.computeIfAbsent(metricGroup, dummy -> new HashMap<>())
+                        .put(metricName, metric);
+            }
         }
     }
 
     @Override
     public void notifyOfRemovedMetric(Metric metric, String metricName, MetricGroup group) {
-        synchronized (this) {
-            removedGroups.add(unwrap(group));
+        if (captureMode != CaptureMode.NONE) {
+            synchronized (this) {
```

Comment:
Nit: Why not make only `removeMetric` synchronized and the `removedGroups` some thread-safe set?

Reply:
Same comment as below. I'd like to keep the threading model as simple as possible for now. See for a reason below.